Dev refactor login

FrauDeborah opened the pull request:
Hi Johnathan (@haxwell)

This PR is NOT yet ready to merge. I need to add the modal unit tests and cleanup the comments/console.logs/status  but should have it complete by Sunday.

Please take a look at the promises to see if I coded them as you had suggested. The promises interaction seam to work, but I would like a sanity check. Files and methods to review are;

- LoginComponent method login()
- LoginService method requestUserLogin()

Feedback is appreciated!

Thank you,

Deborah

FrauDeborah committed:
devFLGetAddCurrentUser branch for saving and getting user in login service. Draft comments, console.logs and homepage button need to be removed before finalizing

FrauDeborah committed:
Merge branch 'dev' into devFLGetAddCurrentUser

FrauDeborah committed:
added UserServie and test files, moved setCurrentUser() and getCurrentUser() to UserService.

FrauDeborah committed:
Removed a few comments and replaced prepareToTransferToRegistation() replaced with cancelLogin(). Worked in chrome

FrauDeborah committed:
merged dev into devRefactorLogin which included PR58. All tests successful. Quizki worked

FrauDeborah committed:
Created loginDialogService and Removed LoginDialogHostComponent. popup works. ng test no failures.

FrauDeborah committed:
devRefactorLogin: added promise for login. ng test no failures.commiting before cleanup and adding more tests

FrauDeborah committed:
added modal tests to loginDialogService

FrauDeborah committed:
added tests for LoginDialogService, LoginComponent, and LoginService; removed console.log button on home page; modified login status; removed comments.

FrauDeborah committed:
merged dev into branch. still has the random setAttribute error

FrauDeborah commented:
Hi Johnathan,

I think the refactoring of the login is finished. So that means perhaps issues #8, #41, #51,  and  #56 can all be closed. 

- LoginDialogHostComponent and associated tests were removed
- LoginDialogService was created to launch the modal for login. 
- LoginComponent was modified to use a promise rather than an observable subscription, loginStatus now has only two messages, commented out code was removed, setCurrentUser was moved to LoginService.
- LoginService was modeifed to use a promise, setCurrentUser added in the promise. 
- UserService was added. setCurrentUser and get CurrentUsers were moved into UserServices
- unit tests were add and all worked until I merge dev into the branch. Then the random setAttribute error surfaced again.

I may need to add one or two more tests. I think the promise between LoginComponent.login() and LoginService.requestUserLogin() should have a test. I need to think about this test. I think the test can be added when I work on the other two issues related to Login. Those 2 new issues are #33 and #35.

Let me know what you think, or merge to dev, if you choose.

Deborah

haxwell commented on src/app/login/login.service.ts:
```diff
@@ -1,22 +1,37 @@
 
-    return this.apiService.get(apiVerifyCredentialsUrl, username, password);
-
+    let promise = new Promise((resolve, reject) => {
```
Yes!!

haxwell commented:
Deborah, this looks great! Thank you for your work!

I do get one error when running the tests however..

```
Chrome 70.0.3538 (Linux 0.0.0) LoginComponent should validate username field value contains at least 5 characters FAILED
	Uncaught TypeError: Cannot read property 'setAttribute' of undefined thrown
```

Do you also get this error?

If you do, perhaps we can just remove this test for now, and get your changes merged. Tests should be green then, ideally. Then we can work on this failing test.

FrauDeborah commented:
Thank you, Johnathan (@haxwell) 

That random Raskie Rabbit of “Uncaught TypeError: Cannot read property 'setAttribute' of undefined thrown”  throws errors on various unpredictable and random tests.  Yet, mostly on LoginComponent tests.

To write tests and to test I used a focused describe, before I merged in the updated dev. All tests passed using focused tests both before and after the dev merge. After removing the focused from describe the setAttribute issue returned. Thus, I don’t think the test which you mentioned is the issue.

I’m going to research to see if I can find a permanent solution for the setAttribute issue. I wasn’t worried about it earlier because I thought it was mentioned at the recent F2F that tinymce was causing the random setAttribute error. Apparently not! I’ll rerun the test to see the error you mentioned.

I did read a StackOverflow article last night about the setAttibute error, which mentioned an Angular issue which is avoided with moving something from a md-menu to a md-input. LoginComponent has no menu and ‘md’ is Angular 5 or less, while we are using Angular 7 with ‘mat’ (Materials) as in mat-input or mat-menu. Thus, I ignored the article. Yet I’ll go back to re-read that article perhaps I can extrapolate a fix for the setAttribute issue.

Thanks for the feedback.

Regards,

Deborah 


On Mar 4, 2019, at 6:50 AM, Johnathan James <notifications@github.com> wrote:

Deborah, this looks great! Thank you for your work!

I do get one error when running the tests however..

Chrome 70.0.3538 (Linux 0.0.0) LoginComponent should validate username field value contains at least 5 characters FAILED
	Uncaught TypeError: Cannot read property 'setAttribute' of undefined thrown
Do you also get this error?

If you do, perhaps we can just remove this test for now, and get your changes merged. Tests should be green then, ideally. Then we can work on this failing test.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

FrauDeborah commented:
Hi Johnathon ( @haxwell ),

Just thought that I would show you how random this issue is. The screenshot below shows that the test failed in a different area for me, than for you. "FAILED LoginComponent should create
debug.js:21 Uncaught TypeError: Cannot read property 'setAttribute' of undefined thrown"
 
![image](https://user-images.githubusercontent.com/23142096/53753457-8e864680-3e6e-11e9-8e82-f47c6fb3e9af.png)

Now this test shows the results with focused describe (fdescribe) on the LoginComponent tests. Notice all 9 LoginComponest tests pass.

![image](https://user-images.githubusercontent.com/23142096/53753788-592e2880-3e6f-11e9-8777-28dcb8e93357.png)


When reading the report in the GitBash window, the report showed fit and fdescribe are deprecated with Jasmine 3.0. Here is the actual report.

```
Chrome 72.0.3626 (Windows 7.0.0) ERROR: 'DEPRECATION: fit and fdescribe will cause your suite to report an 'incomplete' status in Jasmine 3.0'

Chrome 72.0.3626 (Windows 7.0.0) ERROR: 'DEPRECATION: fit and fdescribe will cause your suite to report an 'incomplete' status in Jasmine 3.0'
Chrome 72.0.3626 (Windows 7.0.0): Executed 9 of 62 (skipped 53) SUCCESS (3.685 secs / 4.458 secs)
TOTAL: 9 SUCCESS
TOTAL: 9 SUCCESS
```

Perhaps that is "the why" for the 9 tests passing. Although in package.json the devDependencies show we are not using Jasmine 3.0. Yet in all practicality we are using most of Jasmine 3.0, as jasmine-core is 2.99.1 Maybe we should not use fit or fdescribe.  

```
"devDependencies": {
    "@angular-devkit/build-angular": "^0.12.1",
    "@angular/cli": "^7.1.3",
    "@angular/compiler-cli": "^7.1.3",
    "@angular/language-service": "^7.1.3",
    "@types/jasmine": "^2.8.9",
    "@types/jasminewd2": "^2.0.5",
    "@types/node": "^10.11.7",
    "codelyzer": "~4.2.1",
    "jasmine-core": "~2.99.1",
    "jasmine-spec-reporter": "~4.2.1",
    "karma": "^3.1.4",
    "karma-chrome-launcher": "~2.2.0",
    "karma-coverage-istanbul-reporter": "^1.4.3",
    "karma-jasmine": "~1.1.1",
    "karma-jasmine-html-reporter": "^0.2.2",
    "protractor": "^5.4.1",
    "ts-node": "~5.0.1",
    "tslint": "~5.9.1",
    "typescript": "~3.1.6"
  }
```

Hitting debug button on the test Chrome browser also show deprecation in the inspect console window.

![image](https://user-images.githubusercontent.com/23142096/53754687-eb373080-3e71-11e9-8d76-c0518be4b158.png)

I'll continue my research.

Deborah

FrauDeborah committed:
Hopefully resolved the setAttibute issue by

removing unneeded css tags.

FrauDeborah committed:
fixed ngZone.run warnings in
app-routing module to see if that is causing setAttribute issue. It was not

FrauDeborah committed:
minor updates to login component and test spec to attempt to fix setAttribute issue. Still random problem.

FrauDeborah commented:
Hi Johnathan (@haxwell),

Hoping all is well with your family!  I hate to bother you. Yet, would like some feedback if possible.

I'm still trying to fix the random **setAttribute** issue. (i.e., `Uncaught TypeError: Cannot read property setAttribute of undefined thrown`)

I pushed some changes tonight. Cleaned up some code in the `LoginComponent` and associated tests, and fixed the `ngZone.run()` warning in app-routing tests on the long shot that not using the `ngZone.run()` was causing the issue. Apparently not, because the error still randomly occurs on various tests normally the `LoginComponent`, the `LoginDialogService`, or the `LoginService`.

When I inspect the test results using debug, I can't help wondering if the randomness of where the setAttribute error occurs could be associated with an  async error processed by an earlier test which happens to complete and return the error after the test completes. If an error returned from an earlier test could impact the test which is executing at the exact moment the error returned, then that may explain the randomness of which test fails.  Here is the window sample.  

![image](https://user-images.githubusercontent.com/23142096/54174114-7a33e200-444a-11e9-9510-92f58129ba8b.png)

Notice that `tinymce` is mentioned in the first  `Uncaught TypeError: Cannot read property setAttribute of undefined` message. That first message is presented immediately before the `FAILED LoginCompoent should create`, which is directly before the 2nd error message `Uncaught TypeError: Cannot read property setAttribute of undefined thrown`. Yet, the only components which use tinymce are tested earlier (notice the 2 highlighted lines) a few lined above the 1st error message.

I started to create a spy for the editor in `CreateQuestionPageEditorSectionComponent ` and/or the `CreateQuestionPageComponent` tests, but I didn't quite understand how to create a spy on a script. The  script (at least I think it is a script)  is in the html. I think the issue might be related to the <editior></editor> code below.

```
<mat-card style="width:92%; margin:0 auto;">
    <div class="typeConfig1">Part 2: &nbsp;&nbsp;Enter question</div><br>
    <div id="tinyEditor" style="width:92%; margin:0px auto;">
        <editor apiKey="lzlkf4t1qvjtid356kafllkafx5quoci0htkh5pxp41ilf73" 
            [init]="{
                forced_root_block:'div',
                statusbar: false,
                menubar: false,
                plugins: 'image link hr charmap lists',
                height: 150,
                toolbar: 'undo redo | bold italic underline strikethrough | alignleft aligncenter alignjustify | styleselect | bullist numlist | outdent indent | image | hr removeformat | subscript superscript | charmap '
            }" 
            (onBlur)="submitTextEvent()"
            [(ngModel)]="createQuestionEditorText"
            >
        </editor>
    </div> 
</mat-card>    
```
Also after all the tests complete (success or fail), 4 copies of this error are displayed. which supports the idea that the `cloud.tinymce.com...` was called 4  times before which the test actually completed rendering the errors.

`zone.js:3272 Access to XMLHttpRequest at 'https://cloud.tinymce.com/stable/themes/modern/theme.min.js' from origin 'http://localhost:9876' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

Was wondering if you would comment on my theory and suggest how to handle a spy for the editor?

Deborah

haxwell commented:
Hi Deborah...

I think you are on the right track.  It seems there is a bug in TinyMCE, in
that if it is loaded multiple times (like a test might do, when testing
more than one thing, or one time), then it can produce this error. This
StackOverflow talks about how to correct it, using JQuery, etc..
https://stackoverflow.com/questions/29156329/tinymce-cannot-read-property-setattribute-of-null.
..

Perhaps an entry in the 'providers' section of the TestBed, which supplied
a mock object.. See line 12 of login.service.spec.ts. It provides a jasmine
spy as the value for ApiService. I think we need to do something like that
for TinyMCE.. because we don't really want to test TinyMCE proper.. it
shouldn't be in our test at all.. If we were doing a true end-to-end test,
we would like to see it loaded, and be able to type in it, get the expected
values from it, etc. But for a unit test like this, I don't think we need
it.

Thats one possibility. Another is take out TinyMCE, and use another
Javascript WYSIWYG editor..

Another possibility, is just recognize this is an issue, and we come back
and fix it later.

You've created a great write up.. detailed, well explained, all that.. good
job.

Johnathan

On Mon, Mar 11, 2019 at 11:04 PM Deborah Frauenfelder <
notifications@github.com> wrote:

> Hi Johnathan (@haxwell <https://github.com/haxwell>),
>
> Hoping all is well with your family! I hate to bother you. Yet, would like
> some feedback if possible.
>
> I'm still trying to fix the random *setAttribute* issue. (i.e., Uncaught
> TypeError: Cannot read property setAttribute of undefined thrown)
>
> I pushed some changes tonight. Cleaned up some code in the LoginComponent
> and associated tests, and fixed the ngZone.run() warning in app-routing
> tests on the long shot that note using the ngZone.run() was causing the
> issue. Apparently not, because is the error still randomly occurs on
> various tests normally the LoginComponent, the LoginDialogService, or the
> LoginService.
>
> When I inspect the test results using debug, I can't help wondering if the
> randomness of where the setAttribute error occurs could be associated with
> an async error processed by an earlier test which happens to complete and
> return the error after the test completes. If an error returned from an
> earlier test could impact the test which is executing at the exact moment
> the error returned, then that may explain the randomness of which test
> fails. Here is the window sample.
>
> [image: image]
> <https://user-images.githubusercontent.com/23142096/54174114-7a33e200-444a-11e9-9510-92f58129ba8b.png>
>
> Notice that tinymce is mentioned in the first Uncaught TypeError: Cannot
> read property setAttribute of undefined message. That first message is
> presented immediately before the FAILED LoginCompoent should create,
> which is directly before the 2nd error message Uncaught TypeError: Cannot
> read property setAttribute of undefined thrown. Yet, the only components
> which use tinymce are tested earlier (notice the 2 highlighted lines) a few
> lined above the 1st error message.
>
> I started to create a spy for the editor in CreateQuestionPageEditorSectionComponent
> and/or the CreateQuestionPageComponent tests, but I didn't quite
> understand how to create a spy on a script. The script (at least I think it
> is a script) is in the html. I think the issue might be related to the code
> below.
>
> <mat-card style="width:92%; margin:0 auto;">
>     <div class="typeConfig1">Part 2: &nbsp;&nbsp;Enter question</div><br>
>     <div id="tinyEditor" style="width:92%; margin:0px auto;">
>         <editor apiKey="lzlkf4t1qvjtid356kafllkafx5quoci0htkh5pxp41ilf73"
>             [init]="{
>                 forced_root_block:'div',
>                 statusbar: false,
>                 menubar: false,
>                 plugins: 'image link hr charmap lists',
>                 height: 150,
>                 toolbar: 'undo redo | bold italic underline strikethrough | alignleft aligncenter alignjustify | styleselect | bullist numlist | outdent indent | image | hr removeformat | subscript superscript | charmap '
>             }"
>             (onBlur)="submitTextEvent()"
>             [(ngModel)]="createQuestionEditorText"
>             >
>         </editor>
>     </div>
> </mat-card>
>
> Also after all the tests complete (success or fail), 4 copies of this
> error are displayed. which supports the idea that the cloud.tinymce.com...
> was called 4 times before which the test actually completed rendering the
> errors.
>
> zone.js:3272 Access to XMLHttpRequest at '
> https://cloud.tinymce.com/stable/themes/modern/theme.min.js' from origin '
> http://localhost:9876' has been blocked by CORS policy: No
> 'Access-Control-Allow-Origin' header is present on the requested resource.
>
> Was wondering if you would comment on my theory and suggest how to handle
> a spy for the editor?
>
> Deborah
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/team-quizki/quizki-frontend/pull/62#issuecomment-471858502>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJOzNc73BLLkmfNuknyCPK7H-pNgyw-ks5vVzVhgaJpZM4bVkan>
> .
>

FrauDeborah commented:
Johnathan ( @haxwell ),

To get past the tinymce issues, I marked both tests for
create-question-page and create-question-page-editor-section as pending as
in xdescribe.

Wallah! No errors in the testing debug window console  ... no not one error
appeared!  Hallelujah!!!!

We do still have the initial deprecation warning at the top of the console:
'[Deprecation] Synchronous XMLHttpRequest on the main thread is deprecated
because of its detrimental effects to the end user's experience. For more
help, check https://xhr.spec.whatwg.org/.' I don't see were quizki is using
XMLHttpRequest. Therefore, I presume it a warning for the test environment.

I'll push the final change with the 2 xdescribe tests, then add an issue
with the findings and details. Someone can work the issue later. After all,
you and Brian had discussed changing tinymce anyway. Once tinymce is
changed out the tests can be revisited.

Thank you for the feed back.



Regards,

Deborah Frauenfelder


On Tue, Mar 12, 2019 at 7:17 AM Johnathan James <notifications@github.com>
wrote:

> Hi Deborah...
>
> I think you are on the right track. It seems there is a bug in TinyMCE, in
> that if it is loaded multiple times (like a test might do, when testing
> more than one thing, or one time), then it can produce this error. This
> StackOverflow talks about how to correct it, using JQuery, etc..
>
> https://stackoverflow.com/questions/29156329/tinymce-cannot-read-property-setattribute-of-null
> .
> ..
>
> Perhaps an entry in the 'providers' section of the TestBed, which supplied
> a mock object.. See line 12 of login.service.spec.ts. It provides a jasmine
> spy as the value for ApiService. I think we need to do something like that
> for TinyMCE.. because we don't really want to test TinyMCE proper.. it
> shouldn't be in our test at all.. If we were doing a true end-to-end test,
> we would like to see it loaded, and be able to type in it, get the expected
> values from it, etc. But for a unit test like this, I don't think we need
> it.
>
> Thats one possibility. Another is take out TinyMCE, and use another
> Javascript WYSIWYG editor..
>
> Another possibility, is just recognize this is an issue, and we come back
> and fix it later.
>
> You've created a great write up.. detailed, well explained, all that.. good
> job.
>
> Johnathan
>
> On Mon, Mar 11, 2019 at 11:04 PM Deborah Frauenfelder <
> notifications@github.com> wrote:
>
> > Hi Johnathan (@haxwell <https://github.com/haxwell>),
> >
> > Hoping all is well with your family! I hate to bother you. Yet, would
> like
> > some feedback if possible.
> >
> > I'm still trying to fix the random *setAttribute* issue. (i.e., Uncaught
> > TypeError: Cannot read property setAttribute of undefined thrown)
> >
> > I pushed some changes tonight. Cleaned up some code in the LoginComponent
> > and associated tests, and fixed the ngZone.run() warning in app-routing
> > tests on the long shot that note using the ngZone.run() was causing the
> > issue. Apparently not, because is the error still randomly occurs on
> > various tests normally the LoginComponent, the LoginDialogService, or the
> > LoginService.
> >
> > When I inspect the test results using debug, I can't help wondering if
> the
> > randomness of where the setAttribute error occurs could be associated
> with
> > an async error processed by an earlier test which happens to complete and
> > return the error after the test completes. If an error returned from an
> > earlier test could impact the test which is executing at the exact moment
> > the error returned, then that may explain the randomness of which test
> > fails. Here is the window sample.
> >
> > [image: image]
> > <
> https://user-images.githubusercontent.com/23142096/54174114-7a33e200-444a-11e9-9510-92f58129ba8b.png
> >
> >
> > Notice that tinymce is mentioned in the first Uncaught TypeError: Cannot
> > read property setAttribute of undefined message. That first message is
> > presented immediately before the FAILED LoginCompoent should create,
> > which is directly before the 2nd error message Uncaught TypeError: Cannot
> > read property setAttribute of undefined thrown. Yet, the only components
> > which use tinymce are tested earlier (notice the 2 highlighted lines) a
> few
> > lined above the 1st error message.
> >
> > I started to create a spy for the editor in
> CreateQuestionPageEditorSectionComponent
> > and/or the CreateQuestionPageComponent tests, but I didn't quite
> > understand how to create a spy on a script. The script (at least I think
> it
> > is a script) is in the html. I think the issue might be related to the
> code
> > below.
> >
> > <mat-card style="width:92%; margin:0 auto;">
> > <div class="typeConfig1">Part 2: &nbsp;&nbsp;Enter question</div><br>
> > <div id="tinyEditor" style="width:92%; margin:0px auto;">
> > <editor apiKey="lzlkf4t1qvjtid356kafllkafx5quoci0htkh5pxp41ilf73"
> > [init]="{
> > forced_root_block:'div',
> > statusbar: false,
> > menubar: false,
> > plugins: 'image link hr charmap lists',
> > height: 150,
> > toolbar: 'undo redo | bold italic underline strikethrough | alignleft
> aligncenter alignjustify | styleselect | bullist numlist | outdent indent |
> image | hr removeformat | subscript superscript | charmap '
> > }"
> > (onBlur)="submitTextEvent()"
> > [(ngModel)]="createQuestionEditorText"
> > >
> > </editor>
> > </div>
> > </mat-card>
> >
> > Also after all the tests complete (success or fail), 4 copies of this
> > error are displayed. which supports the idea that the
> cloud.tinymce.com...
> > was called 4 times before which the test actually completed rendering the
> > errors.
> >
> > zone.js:3272 Access to XMLHttpRequest at '
> > https://cloud.tinymce.com/stable/themes/modern/theme.min.js' from
> origin '
> > http://localhost:9876' has been blocked by CORS policy: No
> > 'Access-Control-Allow-Origin' header is present on the requested
> resource.
> >
> > Was wondering if you would comment on my theory and suggest how to handle
> > a spy for the editor?
> >
> > Deborah
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/team-quizki/quizki-frontend/pull/62#issuecomment-471858502
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ADJOzNc73BLLkmfNuknyCPK7H-pNgyw-ks5vVzVhgaJpZM4bVkan
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/team-quizki/quizki-frontend/pull/62#issuecomment-471995967>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AWEe0HvRwf_H5CL31s4SxLB6Vi7ZYjSFks5vV6jggaJpZM4bVkan>
> .
>

FrauDeborah committed:
marked all create-question-page.component tests and create-question-page-editor-section.component tests as pending on the describes. multiple tinymce tests were causing errors. Will create issue on quizki-frontend.

FrauDeborah commented:
Hi Johnathan ( @haxwell ),

Repeating and modifying a previous comment.... 

Please merge PR  devRefactorLogin into team-quizki/quizki-frontend/dev.

I think the refactoring of the login is finished. So that means perhaps issues #8, #41, #51, and #56 can all be closed.

- LoginDialogHostComponent and associated tests were removed
- LoginDialogService was created to launch the modal for login.
- LoginComponent was modified to use a promise rather than an observable subscription, loginStatus now has only two messages, commented out code was removed, setCurrentUser was moved to LoginService.
- LoginService was modeifed to use a promise, setCurrentUser added in the promise.
- UserService was added. setCurrentUser and get CurrentUsers were moved into UserServices
- unit tests were add and all worked until I merge dev into the branch. Then the random setAttribute error surfaced again. In the process of trying to resolve the random issue  I cleaned up some residual and unneeded code, fix some login component tests errors, fixed the ngzone.run() warnings on the app-routing.module hoping that was the issue, and (finally due to tinymce causing the setAttribue error) marked both tests for create-question-page and create-question-page-editor-section as pending as in xdescribe. 

I will create a quizki-frontend issue for tinymce on the  create-question-page/reate-question-page-editor-section components.

I (or someone else) can work on the new login issues #33 and #35 later. Before that point, I need to finish the registration tests to complete that open issue, and I think the issue for  removing the legacy questions code should take a priority over #33 and #35. The longer we wait to extract the questions code, the harder it will be to remove the code.

Thank you for your help. Please merge this PR to dev.

Deborah
